src/DataContainer/EventHelper.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace Contao\CoreBundle\DataContainer;
+
+use Contao\CoreBundle\Framework\ContaoFrameworkInterface;
+use Contao\System;
+
+class EventHelper
+{
+    /**
+     * @var System
+     */
+    private $system;
+
+    /**
+     * Constructor.
+     *
+     * @param ContaoFrameworkInterface $framework
+     */
+    public function __construct(ContaoFrameworkInterface $framework)
+    {
+        $this->system = $framework->getAdapter(System::class);
+    }
+
+    /**
+     * Triggers a callback whilte handling singleton and service classes.
+     *
+     * @param array|callable $callback
+     * @param array          $arguments
+     *
+     * @return mixed
+     */
+    public function trigger($callback, array $arguments = [])
+    {
+        if (\is_array($callback)) {
+            return call_user_func_array(
+                [$this->system->importStatic($callback[0]), $callback[1]],
+                $arguments
+            );
+        }
+
+        if (\is_callable($callback)) {
+            return call_user_func_array($callback, $arguments);
+        }
+
+        throw new \InvalidArgumentException(sprintf('Cannot trigger "%s"', var_export($callback, true)));
+    }
+
+    /**
+     * Triggers an array of callback.
+     *
+     * @param array         $callbacks
+     * @param array         $arguments
+     * @param mixed|null    $returnArgument
+     * @param callable|null $returnIf
+     *
+     * @return mixed
+     */
+    public function triggerAll(array $callbacks, array $arguments = [], $returnArgument = null, callable $returnIf = null)
```

Comment:
I like this in general. The ``$returnArgument`` index looks a bit unintuitive to me however. Maybe use a callback instead to select the input for the next round or even merge with the ``$returnIf`` one?

Reply:
How would a callback work?

Reply:
Something like this:
```PHP
$arguments = $selectorFunction($arguments, $result);
```

Which you could consume like so:
```PHP
triggerAll(
    [$callback1, $callback2, $callback3, $callback4]
    [$a, $b, $c],
    function($args, $result) { $args[0] = true; $args[1] = $result; return $args;  }
);
```

Or directly asking for arguments each run (then there could be an exception thrown to abort):
```PHP
$args = [ $enabled, $this, $that ];
triggerAll(
    [$callback1, $callback2, $callback3, $callback4]
    function($index, $previousResult) use $args { 
        if($index > 0) { $args[1] = $previousResult; }
        if('dangerous' === $previousResult) { throw new StopPropagationException(); }
        return $args;
    }
);
```
But that's not that elegant as well to be honest. It all depends on the primary usecase. If that's piping something through all callbacks, then your approach is probably the best. But maybe consider naming the parameter `returnArgumentIndex` or something to make clear that it relates to `arguments`?

Reply:
`$args` is an array, so it will not be passed by reference (variant 2 won't work). Variant 1 *might*, if you pass `$args` as reference (`&$args`). Haven't done that with a closure though.

Reply:
I'm not sure if we mean the same thing? https://3v4l.org/Z0Fmh

Reply:
The example does not qualify. Contrary to your code, the `triggerAll` method in the implementation does not live within the same class, so you can't `use` the same variable.

Reply:
The use statement makes a closure that binds & extends lifetime of the local variables. It's the same as putting the `$args` in my example inside the function and omitting the `use`.

Reply:
You can't `use` `$args` because it does not exist in the place where you create the closure 😉

Reply:
Sure, the closure gets created on the consumer/caller side where the args are available (and it would work without `use` if you put `$args` directly into the anonymous function). So I'm still not sure if we are talking about the same thing. But I think this discussion misses the point anyways. :grinning:

Reply:
Here's an example: https://3v4l.org/ZCNdN

1. passing `$args` by reference does not work in closures
2. `use`ing the original `$arguments` does not work by default
3. `use`ing the original `$arguments` by reference allows you to modify the original arguments, but that does not pass them on the multiple callbacks
4. The only thing that _does_ work is if your closure returns the arguments array that should be used for the successive call.